src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -272,9 +259,182 @@ void RuntimeConfigurator::update_data_offsets() const {
     }
 }
 
-void RuntimeConfigurator::update_latest_shapes() {
-    for (size_t i = 0; i < m_io_num; ++i) {
-        m_latest_shapes[i] = m_io_descs[i]->get_shape();
+std::vector<VectorDims> RuntimeConfigurator::extract_shapes() {
+    std::vector<VectorDims> shapes(m_io_num);
+    for (size_t i = 0; i < m_io_num; ++i)
+        shapes[i] = m_io_descs[i]->get_shape();
+    return shapes;
+}
+
+std::vector<std::vector<size_t>> RuntimeConfigurator::extract_layouts() {
+    std::vector<std::vector<size_t>> layouts(m_io_num);
+    for (size_t i = 0; i < m_io_num; ++i)
+        layouts[i] = m_io_descs[i]->get_layout();
+    return layouts;
+}
+
+RuntimeConfigurator::UnifiedLoopInfoRtParams RuntimeConfigurator::compute_runtime_params(const lowered::UnifiedLoopInfoPtr& loop_info) {
+    RuntimeConfigurator::UnifiedLoopInfoRtParams rt_params;
+    rt_params.work_amount = loop_info->get_work_amount();
+    const auto count = loop_info->get_input_count() + loop_info->get_output_count();
+    rt_params.ptr_increments.resize(count);
+    rt_params.finalization_offsets.resize(count);
+
+    size_t idx = 0;
+    loop_info->iterate_through_descs(
+        [&rt_params, &idx](const lowered::UnifiedLoopInfo::LoopPortDesc& desc) {
+            rt_params.ptr_increments[idx] = desc.ptr_increment;
+            rt_params.finalization_offsets[idx] = desc.finalization_offset;
+            ++idx;
+        });
+    return rt_params;
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::init(const lowered::LinearIRPtr& linear_ir,
+                                                    const std::vector<PortDescriptorPtr>& io_descs,
+                                                    size_t in_num) {
+    if (linear_ir->get_config().m_enable_domain_optimization || !linear_ir->is_dynamic())
+        return;
+    concurrency = linear_ir->get_config().m_min_parallel_work_amount;
+
+    const auto brgemms = find_applicable_brgemms(linear_ir);
+    // Parallel WA optimization is Brgemm related
+    if (brgemms.empty())
+        return;
+
+    init_non_m_related_params(linear_ir, brgemms);
+    OPENVINO_ASSERT(!not_m_related_params.empty(), "not_m_related_params mustn't be empty after initialization");
```

Comment:
Probably it would be better:
```
// At the moment this optimization is Brgemm related so there must be `unsqueezed_params`
m_unsqueezed_params = init_unsqueezed_params(linear_ir); // <- `find_applicable_brgemms()` is called inside
OPENVINO_ASSERT(!m_unsqueezed_params .empty(), "m_unsqueezed_params mustn't be empty after initialization");
.....
```

What do you think?

Reply:
Actually we must check that `brgemms` is not empty before initialization (in order to skip non brgemm related Subgraphs), so we can't call it inside `init_unsqueezed_params`

However, I think you are right, and it's better to have return value in these initialization methods, I refactored it a bit. Could you please take a look?